In [0]:
from __future__ import print_function
import numpy as np

In [0]:
np.zeros(10)

array([0., 0., 0., 0., 0., 0., 0., 0., 0., 0.])

In [0]:
line1=[6,7,8,9,10]
arr1=np.array(line1)
print (arr1)

[ 6  7  8  9 10]


In [0]:
np.zeros((3,6))

array([[0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0.]])

In [0]:
np.ones(10)

array([1., 1., 1., 1., 1., 1., 1., 1., 1., 1.])

In [0]:
l=[6,7,8,9,10]
narr=np.array(l)
narr.dtype

dtype('int64')

In [0]:
array2=np.random.rand(2,4)

In [0]:
array2.dtype # the following function will give the dimention of the array

dtype('float64')

In [0]:
array2.ndim # the following function will give the dimension in (row, column)

2

In [0]:
array2.shape

(2, 4)

#reshaping

In [0]:
array1=np.arange(10, dtype=float)
print(array1)

[0. 1. 2. 3. 4. 5. 6. 7. 8. 9.]


In [0]:
array1=np.arange(10, dtype=float).reshape(2,5) 
# reshape array to 2 rows and 5 columns
print(array1) 
print(array1.shape)

[[0. 1. 2. 3. 4.]
 [5. 6. 7. 8. 9.]]
(2, 5)


In [0]:
print(array1.reshape(5,2))

[[0. 1.]
 [2. 3.]
 [4. 5.]
 [6. 7.]
 [8. 9.]]


In [0]:
#add ann axis
b=np.array([0,1])
b_col=b[:,np.newaxis]
print(b)
print(b_col)
# change the axix of array: rows into columns

[0 1]
[[0]
 [1]]


In [0]:
#transpose
print(b)
print(b_col.T)
# returns a view of the array with axes transposed

[0 1]
[[0 1]]


In [0]:
# flatten: always returns a flat copy of the original array
a = np.array([[1,2], [3,4]])
a.flatten()

a.flatten('F')


array([1, 3, 2, 4])

In [0]:
array1=np.arange(10, dtype=float).reshape(2,5)
array1_flt=array1.flatten()
array1_flt[4]=13
print(array1)
print(array1_flt)

[[0. 1. 2. 3. 4.]
 [5. 6. 7. 8. 9.]]
[ 0.  1.  2.  3. 13.  5.  6.  7.  8.  9.]


In [0]:
array1=np.arange(10, dtype=float).reshape(2,5)
array1_flt=array1.ravel()
array1_flt[4]=17

print(array1_flt)
print(array1)

# revel returns a contiguous flattened array

[ 0.  1.  2.  3. 17.  5.  6.  7.  8.  9.]
[[ 0.  1.  2.  3. 17.]
 [ 5.  6.  7.  8.  9.]]


# Task 4.4 Stack arrays

In [0]:

# stack flat arrays in columns
a = np.array([0,1])
b = np.array([2,3])

# stack will put the two arrays in column , stack on top of eachother
ab = np.stack((a,b)).T
print(ab)

# Another way of doing it

np.hstack((a[:, None], b[:,None]))

[[0 2]
 [1 3]]


array([[0, 2],
       [1, 3]])

#4.5 Selection 

In [0]:

#single item

arr = np.arange(10, dtype=float).reshape((2,5))
print(arr)


arr[0] # This will return the first element of the array
arr[0,3] # it return the element of 0 row and column 3 (the row and column numbered from zero)
# alternate method of doing the same thing as above
arr[0][3]


[[0. 1. 2. 3. 4.]
 [5. 6. 7. 8. 9.]]


3.0

# 4.5.1 Slicing

In [0]:

#Syntax:start:stop: step with start (default 0) stop (default 1)

#row 0: returns 1d array 

arr[0, :] 

# return an array consists of elements from "0 column" (1st count)
arr[:, 0]

# return an array consists of elements from "2 column" (3rd count)
arr[:,2]


# return an array from "2 column" (3rd count) onwards
arr[:,2:]
print(arr)
print(arr[:,2:])

# return an array from "1 column" (2ndcount) onwards
''' array1 =
[[1 2 3 4]
 [5 6 7 8]]'''
array2 = arr[: , 1:4]
array2


[[0. 1. 2. 3. 4.]
 [5. 6. 7. 8. 9.]]
[[2. 3. 4.]
 [7. 8. 9.]]


array([[1., 2., 3.],
       [6., 7., 8.]])

# fancy indexing: integer or boolean array indexing


In [0]:
# fancy indexing returns a copy not a view !!!!

arr=np.arange(10, dtype=float).reshape(2,5)
array2=arr[:,[1,2,3]] 
print(arr)
print(array2)

[[0. 1. 2. 3. 4.]
 [5. 6. 7. 8. 9.]]
[[1. 2. 3.]
 [6. 7. 8.]]


In [0]:
array2[0,2]=20
print(array2)

[[44.  2. 20.]
 [ 6.  7.  8.]]


In [0]:
array2=arr[arr>5]
print(array2)

[6. 7. 8. 9.]


In [0]:
array2[0]=33
print(array2)

[33.  7.  8.  9.]


In [0]:
print(arr)

[[0. 1. 2. 3. 4.]
 [5. 6. 7. 8. 9.]]


In [0]:
# however, in the context of value indexing 'Fancy' authorizes
# the modification of yhe original array

arr[arr>5]=0
print(arr)

[[0. 1. 2. 3. 4.]
 [5. 0. 0. 0. 0.]]


In [0]:
# booleans arrays indexing continoues

fruits=np.array(['orange', 'banana','mango','apple'])
fruits=='mango'

array([False, False,  True, False])

In [0]:
fruits[fruits!='banana']

array(['orange', 'mango', 'apple'], dtype='<U6')

In [0]:
(fruits=='banana')|(fruits=='ananas')

array([False,  True, False, False])

In [0]:
fruits[fruits!='banana']='peach'
print(fruits)

['peach' 'banana' 'peach' 'peach']


In [0]:
np.unique(fruits)

array(['banana', 'peach'], dtype='<U6')

4.6

# 4.6 Vectorized Operations

In [0]:


nums = np.arange(5)
nums * 10                # multiply each by 10
nums  = np.sqrt (nums) # square root of each element
np.ceil(nums)         # also floor, rint (round the nearest int)
np.isnan(nums)       # checks for NaN
nums +np.arange(5)   # add element -wise
np.maximum(nums, np.array([1, -2, 3, -4, 5])) # compare elemnt -wise




# Compare Euclidean distance between 2 vectors

In [0]:
vec1 = np.random.randn(10) # creating an array of random floats
vec2 = np.random.randn(10) 
dist = np.sqrt(np.sum((vec1-vec2)**2))
print(dist)


# math and stats

In [0]:
rnd = np.random.randn(4, 2) # creating a random array 4X2
rnd.mean()
rnd.std()
rnd.argmin()         # index of minimum element
rnd.sum()
rnd.sum(axis=0)      # sum of columns
rnd.sum(axis=1)     # sum of rows


# methods for boolean arrays



In [0]:
(rnd > 0).sum()         # counts number of positive values
(rnd > 0).any()         # checks if any value its True
(rnd > 0).all()        # checks if all values are True

# random numbers

In [0]:
np.random.seed(12234)      # set the seed
np.random.rand(2, 3)       # 2 X 3 matrix in [0, 1]
np.random.randn(1)        # random normals (mean 0, sd 1)
np.random.randint(0, 2, 10)  # 10 randomly picked 0 or 1

# broadcasting

In [0]:
c=np.array([[1,1,1],
           [40,40,40],
           [50,50,50],
           [60,60,60]])
d=np.array([4,5,6])

print(c+d)

[[ 5  6  7]
 [44 45 46]
 [54 55 56]
 [64 65 66]]


# Excercises



In [0]:
wek=np.random.rand(4,2)

In [0]:
print(wek)

[[0.2549697  0.49724061]
 [0.60506698 0.97821605]
 [0.9056895  0.63981214]
 [0.41899831 0.79595982]]


In [0]:
np.argmin(wek, axis=0) # returns position

array([0, 0])

In [0]:
np.amin(wek, axis=0) #returns values 

array([0.2549697 , 0.49724061])

In [0]:
result=np.where(wek==np.amin(wek, axis=0)) # other possibility - to get exact position
print (result)

(array([0, 0]), array([0, 1]))


In [0]:
result = np.where(wek == np.amin(wek))
print(result)

(array([0]), array([0]))


# standardize

In [0]:
print(wek)


[[0.2549697  0.49724061]
 [0.60506698 0.97821605]
 [0.9056895  0.63981214]
 [0.41899831 0.79595982]]


In [0]:
sigma=(np.std(wek))
print(sigma)
ni=(wek.mean())
print(ni)
z=(wek-ni)/sigma
print(z)

0.23127982625079335
0.6369941402735797
[[-1.65178451 -0.60426166]
 [-0.13804559  1.47536391]
 [ 1.16177604  0.01218439]
 [-0.94256311  0.68733053]]


In [0]:
z.mean()

1.3877787807814457e-17

In [0]:
def stand():
  sigma=(np.std(wek))
  ni=(wek.mean())
  z=(wek-ni)/sigma
  print(z)

In [0]:
stand()

[[-1.65178451 -0.60426166]
 [-0.13804559  1.47536391]
 [ 1.16177604  0.01218439]
 [-0.94256311  0.68733053]]


# page 7

On the page 7 we have arithmetical operation on arrays. If we want to compare them or add they must to be of the same dimension. If one is 1D, it stretch to fit into the dimention of bigger one. Values of 1D array are used to fill in the rest of "dimensions".

#Lists of cammands that are useful and were not included in the exercise

1. np.indices(dimensions, dtype=<type 'int'>)
Indices()  creates a set of arrays stacked as one higher dimession array, one per dimension with each representing variation in that dimension.Return an array representing the indices of a grid.
Compute an array where the subarrays contain index values 0,1,… varying only along the corresponding axis.
Parameters:	
dimensions : sequence of ints

The shape of the grid.
dtype : dtype, optional
Data type of the result.

Returns:	
grid : ndarray
The array of grid indices, grid.shape = (len(dimensions),) + tuple(dimensions).


2.np.cos(x[, out]) = ufunc ‘cos’) : This mathematical function helps user to calculate trigonmetric cosine for all x (being the array elements)

3.np.round(arr, decimals = 0, out = None) : This mathematical function round an array to the given number of decimals.
For values exactly halfway between rounded decimal values, NumPy rounds to the nearest even value. Thus 1.5 and 2.5 round to 2.0, -0.5 and 0.5 round to 0.0, etc. Results may also be surprising due to the inexact representation of decimal fractions in the IEEE floating point standard and errors introduced when scaling by powers of ten.

4.np.transpose(a,axes=None), it changes the order of an array from (r,c) tp (c,r). It allows to change rows into columns. Useful while cleaning and preparing data for analysis.

Parameters:

a : array_like
Input array.

axes : list of ints, optional

By default, reverse the dimensions, otherwise permute the axes according to the values given.

5.np.divide(x1, x2, /, out=None, *, where=True, casting='same_kind', order='K', dtype=None, subok=True[ , signature, extobj]) it divides one array to another.
Behavior on division by zero can be changed using seterr.

Parameters:	

x1 : array_like
Dividend array.

x2 : array_like
Divisor array.

out : ndarray, None, or tuple of ndarray and None, optional
A location into which the result is stored. If provided, it must have a shape that the inputs broadcast to. If not provided or None, a freshly-allocated array is returned. A tuple (possible only as a keyword argument) must have length equal to the number of outputs.

where : array_like, optional
Values of True indicate to calculate the ufunc at that position, values of False indicate to leave the value in the output alone.

kwargs
